front50-web/src/main/groovy/com/netflix/spinnaker/front50/controllers/V2PipelineTemplateController.java

```diff
@@ -62,29 +74,44 @@
   }
 
   @RequestMapping(value = "", method = RequestMethod.POST)
-  void save(@RequestBody PipelineTemplate pipelineTemplate) {
-    checkForDuplicatePipelineTemplate(pipelineTemplate.getId());
-    getPipelineTemplateDAO().create(pipelineTemplate.getId(), pipelineTemplate);
+  void save(@RequestParam(value = "version", required = false) String version, @RequestBody PipelineTemplate pipelineTemplate) {
+    String templateId = StringUtils.isNotEmpty(version) ? String.format("%s:%s", pipelineTemplate.getId(), version) : pipelineTemplate.getId();
+    checkForDuplicatePipelineTemplate(templateId);
+    getPipelineTemplateDAO().create(templateId, pipelineTemplate);
+    saveDigest(pipelineTemplate);
   }
 
-  @RequestMapping(value = "{id}", method = RequestMethod.GET)
-  PipelineTemplate get(@PathVariable String id) {
-    return getPipelineTemplateDAO().findById(id);
+  private void saveDigest(PipelineTemplate pipelineTemplate) {
+    String digestId = String.format("%s@SHA256%s", pipelineTemplate.getId(), computeSHA256Digest(pipelineTemplate));
+    try {
+      checkForDuplicatePipelineTemplate(digestId);
+    } catch (DuplicateEntityException dee) {
+      log.debug("Duplicate pipeline digest calculated, not updating key {}", digestId);
+      return;
+    }
+    getPipelineTemplateDAO().create(digestId, pipelineTemplate);
   }
 
   @RequestMapping(value = "{id}", method = RequestMethod.PUT)
-  PipelineTemplate update(@PathVariable String id, @RequestBody PipelineTemplate pipelineTemplate) {
-    PipelineTemplate existingPipelineTemplate = getPipelineTemplateDAO().findById(id);
-    if (!pipelineTemplate.getId().equals(existingPipelineTemplate.getId())) {
-      throw new InvalidRequestException("The provided id " + id + " doesn't match the pipeline template id " + pipelineTemplate.getId());
-    }
-
+  PipelineTemplate update(@PathVariable String id, @RequestParam(value = "version", required = false) String version, @RequestBody PipelineTemplate pipelineTemplate) {
+    String templateId = StringUtils.isNotEmpty(version) ? String.format("%s:%s", id, version) : pipelineTemplate.getId();
     pipelineTemplate.setLastModified(System.currentTimeMillis());
-    getPipelineTemplateDAO().update(id, pipelineTemplate);
-
+    getPipelineTemplateDAO().update(templateId, pipelineTemplate);
+    updateDigest(pipelineTemplate);
     return pipelineTemplate;
   }
 
+  private void updateDigest(PipelineTemplate pipelineTemplate) {
+    String digestId = String.format("%s@sha256%s", pipelineTemplate.getId(), computeSHA256Digest(pipelineTemplate));
```

Comment:
in the save digest you use uppercase for sha

Reply:
Should be lower, the key is cast to lowercase no matter what it looks like beforehand.